Initial version of dredd plugin

voldern opened the pull request.

voldern committed:
Initial version of Dredd plugin

dlmr commented on extensions/roc-plugin-dredd/src/actions/dredd.js:
```diff
@@ -0,0 +1,33 @@
+    const port = process.env.PORT || settings.runtime.port;
+
+    const dredd = new Dredd({
+        server: `http://localhost:${port}`,
```
I suggest using the `port` and `path` that the action gets as arguments.

https://github.com/rocjs/roc-package-web-app/blob/master/extensions/roc-package-web-app-dev/docs/Hooks.md#server-started

I would also in this case put all of the three nested callbacks together, something like this.

``` js
export default () => (port, path) => () => {
  const dredd = new Dredd({
```

dlmr commented on extensions/roc-plugin-dredd/src/actions/dredd.js:
```diff
@@ -0,0 +1,33 @@
+import { initLog } from 'roc';
+import Dredd from 'dredd';
+
+const log = initLog();
```
Consider adding the `name` and `version` for the extension in the `initLog` function.

https://github.com/rocjs/roc/blob/master/docs/API.md#initlog

This will give more details to the end user about what extension that created the logged message.

dlmr commented on extensions/roc-plugin-dredd/src/actions/dredd.js:
```diff
@@ -0,0 +1,33 @@
+        }
+
+        if (stats.errors) {
+            log.large.error('One or more errors occured while running dredd');
```
You do not need to do an early return here, `error` level logging will close down the process manually using `process.exit`. If this is not desired `warning` can be used instead.

dlmr commented on extensions/roc-plugin-dredd/src/config/roc.config.js:
```diff
@@ -1 +1,16 @@
+export default {
+    settings: {
+        test: {
+            dredd: {
```
Does it make sense to expose all Dredd options through Roc, mainly based on a signal to noise ratio. Are all settings something that the user of this plugin might want to modify and especially set through using command line options.

Another option to minimise the noise but still keep the full configurable power is to use `__raw`.

https://github.com/rocjs/roc/blob/master/docs/Settings.md#escape-hatch

See related discussion here: https://github.com/rocjs/roc-plugin-browsersync/issues/6

dlmr replied:
Some of the settings would make sense to be able to define on a "global" perspective. Thinking about `level` and `silent`, maybe they eventually can be something that can be defined on the highest level and then used throughout the system. A version of this, although not as detailed, is to change the `level` based on if Roc is started in `verbose` mode or not.

voldern replied:
I've rewritten logging in the plugin to reflect this input. We now only rely on the `verbose` flag. I've removed all options except `path`.

dlmr commented on extensions/roc-plugin-dredd/src/roc/index.js:
```diff
@@ -1,7 +1,32 @@
+            dredd: {
+                command: lazyRequire('../commands/dredd'),
+                description: 'Runs dredd in current project',
+                settings: true,
```
No settings are used in this command and this can be removed.

dlmr commented on extensions/roc-plugin-dredd/src/roc/index.js:
```diff
@@ -1,7 +1,32 @@
+        {
+            hook: 'server-started',
+            description: 'Run dredd tests',
+            action: lazyRequire('../actions/dredd'),
```
With the current setup this action would start at all times when the `server-started` hook is invoked and this might not be desired.

One alternative is to register the action when the `dredd` command is invoked, this would mean that it does not exist when not running that command. As a result it would not be invoked any other times than when `dredd` is used with the downside that Roc will not be able to generate documentation for it. 

An alternative is make the action itself conditional based on the context for example.

dlmr replied:
This action will be invoked each time the server restarts, that will be triggered by default via modifications to the code, the `roc.config` or the `/config` directory. 

Does the action itself manage this currently? Should the current Dredd run be canceled in that case and restart it or should all of those modifications be ignored?

voldern committed:
Initialize log using `name` and `version`

voldern committed:
Remove superflous return statement after error logging

voldern committed:
Remove superflous settings option on dredd command

voldern committed:
Remove a lot of configuration options

Remove a lot of config options that probably won't be used most of the time
and instead support raw options

voldern committed:
Use port and path from action arguments

voldern committed:
Make watch mode work

Make watch mode work by queueing dredd tests

voldern committed:
Only register dredd action when running dredd command

We should only register dredd action when running dredd command.

Ref https://github.com/voldern/roc-plugin-dredd/pull/1#discussion_r85939295

voldern committed:
Output more information about dredd run

voldern committed:
Reduce dredd log level

Do not output as much information from dredd, as we output most of the
information our self.

voldern committed:
Set dredd log level to `debug` when running in verbose mode

voldern committed:
Add watch option to dredd command

Exit after running dredd tests by default and add watch command for
going into watch mode.

voldern committed:
Add support for watch mode

Support watch mode by not aborting in the case of an error when the
dev server is stopping. We also implement our own reporter to handle
that case. Our own reporter also gives the added benefit of being able
to do logging how we want.
